### PR TITLE
TDL-15861 added request timeout and backoff

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Added request timeout and backoff for the timeout exception

# Manual QA steps
 - Unit tests to verify the API calls are retried properly
 - Unit tests to verify that the functions with no backoff are not retried
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 version: 2.1
 orbs:
   slack: circleci/slack@3.4.2
-
```
Comment (dbshah1212) [inline .circleci/config.yml:4]: Please revert this change.

File: tap_bing_ads/__init__.py [hunk 2]
```diff
@@ -98,19 +100,37 @@ def set_options(self, **kwargs):
...
 
+def is_timeout_error():
+    def gen_fn(exc):
+        if 'timed out' in str(exc):
+            return False
+            # retry if the error string contains timed out
+        return True
+    return gen_fn
+
+@backoff.on_exception(backoff.expo,
```
Comment (dbshah1212) [inline tap_bing_ads/__init__.py:115]: Please add needful comments here.

File: tap_bing_ads/__init__.py [hunk 3]
```diff
@@ -556,6 +586,11 @@ def sync_campaigns(client, account_id, selected_streams):
 
         return map(lambda x: x['Id'], campaigns)
 
+@backoff.on_exception(backoff.expo,
+                      urllib.error.URLError,
+                      giveup=is_timeout_error(),
+                      max_tries=5,
+                      factor=2)
 def sync_ad_groups(client, account_id, campaign_ids, selected_streams):
```
Comment (dbshah1212) [inline tap_bing_ads/__init__.py:594]: Please add a comment.

Comment (namrata270998) [reply tap_bing_ads/__init__.py:601]: done

File: tap_bing_ads/__init__.py [hunk 4]
```diff
@@ -98,19 +100,37 @@ def set_options(self, **kwargs):
         self._options = kwargs
         kwargs = ServiceClient._ensemble_header(self.authorization_data, **self._options)
         kwargs['headers']['User-Agent'] = get_user_agent()
+        # setting the timeout parameter using the set_options which sets timeout in the _soap_client
+        kwargs['timeout'] = CONFIG.get('request_timeout', DEFAULT_REQUEST_TIMEOUT)
         self._soap_client.set_options(**kwargs)
 
+def is_timeout_error():
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_bing_ads/__init__.py:107]: done

File: tap_bing_ads/__init__.py [hunk 5]
```diff
@@ -98,8 +127,18 @@ def set_options(self, **kwargs):
         self._options = kwargs
         kwargs = ServiceClient._ensemble_header(self.authorization_data, **self._options)
         kwargs['headers']['User-Agent'] = get_user_agent()
+        request_timeout = CONFIG.get('request_timeout')
+        # if request_timeout is other than 0,"0" or "" then use request_timeout
```
Comment (dbshah1212) [inline tap_bing_ads/__init__.py:131]:
```suggestion
        # if request_timeout is other than 0, "0" or "" then use default request_timeout
```

File: tap_bing_ads/__init__.py [hunk 6]
```diff
@@ -98,8 +127,18 @@ def set_options(self, **kwargs):
...
+        if request_timeout and float(request_timeout):
+            request_timeout = float(request_timeout)
+        else: # If value is 0,"0" or "" then set default to 300 seconds.
```
Comment (dbshah1212) [inline tap_bing_ads/__init__.py:134]:
```suggestion
        else: # If value is 0, "0" or "" then set default to 300 seconds.
```

File: tap_bing_ads/__init__.py [hunk 7]
```diff
@@ -679,13 +727,19 @@ async def poll_report(client, account_id, report_name, start_date, end_date, req
...
                       requests.exceptions.ConnectionError,
                       max_tries=5,
                       on_backoff=log_retry_attempt)
 def stream_report(stream_name, report_name, url, report_time):
     with metrics.http_request_timer('download_report'):
-        response = SESSION.get(url, headers={'User-Agent': get_user_agent()})
+        # Set request timeout with config param `request_timeout`.
+        # If value is 0,"0","" or not passed then set default to 300 seconds.
+        config_request_timeout = CONFIG.get('request_timeout')
+        timeout = config_request_timeout and  float(config_request_timeout) or REQUEST_TIMEOUT
```
Comment (dbshah1212) [inline tap_bing_ads/__init__.py:741]: Use if-else instead of the ternary operation to keep consistency and avoid pylint error in the future.

File: tap_bing_ads/__init__.py [hunk 8]
```diff
@@ -98,19 +100,43 @@ def set_options(self, **kwargs):
         self._options = kwargs
         kwargs = ServiceClient._ensemble_header(self.authorization_data, **self._options)
         kwargs['headers']['User-Agent'] = get_user_agent()
+        # setting the timeout parameter using the set_options which sets timeout in the _soap_client
+        kwargs['timeout'] = float(CONFIG.get('request_timeout', REQUEST_TIMEOUT) or REQUEST_TIMEOUT)
+        LOGGER.info(f'>>>>>> kwargs {kwargs}')
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_bing_ads/__init__.py]: done

File: tap_bing_ads/__init__.py [hunk 9]
```diff
@@ -539,6 +570,12 @@ def sync_accounts_stream(account_ids, catalog_item):
     singer.write_bookmark(STATE, 'accounts', 'last_record', max_accounts_last_modified)
     singer.write_state(STATE)
 
+# backoff when URLError occurs with time out
+@backoff.on_exception(backoff.expo,
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_bing_ads/__init__.py]: done

File: tap_bing_ads/__init__.py (hunk 7 above)
Comment (namrata270998) [reply tap_bing_ads/__init__.py:741]: updated the logic to use if-else instead of ternary operation

File: .circleci/config.yml (hunk 1 above)
Comment (namrata270998) [reply .circleci/config.yml:4]: done

File: .circleci/config.yml [hunk 10]
```diff
@@ -25,7 +35,6 @@ jobs:
             run-test --tap=tap-bing-ads tests
       - slack/notify-on-failure:
           only_for_branches: master
-
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply .circleci/config.yml:28]: done

File: tap_bing_ads/__init__.py (hunk 7 above)
Comment (savan-chovatiya) [reply tap_bing_ads/__init__.py:741]: @namrata270998 It's looks like you missed this change. Can you use if-else to keep consistency?

Comment (namrata270998) [reply tap_bing_ads/__init__.py]: Right, missed it as it was present at 2 places
Updated it

**Review by savan-chovatiya — APPROVED** [review]
